File: data-structures/Python/linked_list.py
class Node():
  def __init__(self, data):
    self.data = data
    self.next_node = None

class LinkedList():
  def __init__(self):
    self.head = None

  def add(self, data):
    # adds a new node to the end of the list
    node = Node(data)
    if (self.head == None):
      # list is empty
      self.head = node
    else:
      current = self.head
      while(current.next_node != None):
        current = current.next_node
      current.next_node = node

  def print_list(self):
    if(self.head != None):
      current = self.head
      print(current.data)
      while(current.next_node != None):
        current = current.next_node
        print(current.data)

  def delete(self, data):
    # find node to delete based on provided data
    if (self.head == None):
      return False
    current = self.head
    if(current.data == data):
      # need to delete the head node
      temp_node = current.next_node
      current.next_node = None
      self.head = temp_node
    else:
      # need to keep looping to find the data
      while (current.next_node != None):
        previous = current
        current = current.next_node
        if (current.data == data):
          temp_node = current.next_node
          previous.next_node = temp_node
          current.next_node = None
          return True
      return False

  def insert_before(self, n, data):
    node = Node(data)
    if(n == 0):
      # insert before the head
      temp_node = self.head
      self.head = node
      self.head.next_node = temp_node
    else:
      current = self.head
      previous = None
      for i in range(0, n):
        previous = current
        current = current.next_node
        if (current == None):
          previous.next_node = node
          return True
      temp_node = previous.next_node
      previous.next_node = node
      node.next_node = temp_node

# Sample linked list 
ll = LinkedList()
ll.add(1)
ll.add(2)
ll.add(3)
ll.print_list()
ll.insert_before(2, 5)
ll.print_list()

File: data-structures/JavaScript/linkedList.js
class Node {
  constructor(data) {
    this.data = data;
    this.nextNode = null;
  }
}

class LinkedList {
  constructor() {
    this.head = null;
  }

  add(data) {
    let node = new Node(data);
    // if list empty, point head at this new node
    if(!this.head) {
      this.head = node;
    } else {
      // traverse list and point last at this new node
      let current = this.head;
      while(current.nextNode) {
        current = current.nextNode;
      }
      current.nextNode = node;
    }
  }

  delete(data) {
    if(!this.head) {
      return false;
    }
    let current = this.head;
    if(current.data === data) {
      let tempNode = current.nextNode;
      current.nextNode = null;
      this.head = tempNode;
    } else {
      while(current.nextNode) {
        let previous = current;
        current = current.nextNode;
        if(current.data === data) {
          let tempNode = current.nextNode;
          previous.nextNode = tempNode;
          current.nextNode = null;
          return true;
        }
      }
      return false;
    }
  }

  insertBefore(n, data) {
    let node = new Node(data);
    if(n === 0) {
      let tempNode = this.head;
      this.head = node;
      this.head.nextNode = tempNode;
    } else {
      let current = this.head;
      let previous = null;
      for(let i = 0; i < n; i++) {
        previous = current;
        current = current.nextNode;
        if(current === null) {
          previous.nextNode = node;
          return true;
        }
      }
      let tempNode = previous.nextNode;
      previous.nextNode = node;
      node.nextNode = tempNode;
    }

  }

  printList() {
    if(this.head) {
      let current = this.head;
      console.log(current.data);
      while(current.nextNode) {
        current = current.nextNode;
        console.log(current.data);
      }
    }
  }
}

const list = new LinkedList();
list.add(1);
list.add(14);
list.add(5);
list.add(7);
list.printList();
list.delete(14);
list.printList();
list.insertBefore(2, 23);
list.printList();
